src/strkey.rs

```diff
@@ -111,17 +127,167 @@ impl FromStr for StrkeyPrivateKeyEd25519 {
+    pub fn to_string(&self) -> String {
+        let mut payload: [u8; 40] = [0; 40];
+        let (ed25519, id) = payload.split_at_mut(32);
+        ed25519.copy_from_slice(&self.ed25519);
+        id.copy_from_slice(&self.id.to_be_bytes());
+        encode(version::MUXED_ACCOUNT_ED25519, &payload)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        let (ed25519, id) = payload.split_at(32);
+        Ok(Self {
+            ed25519: ed25519.try_into().map_err(|_| DecodeError::Invalid)?,
+            id: u64::from_be_bytes(id.try_into().map_err(|_| DecodeError::Invalid)?),
+        })
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::MUXED_ACCOUNT_ED25519 => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeyPreAuthTx(pub [u8; 32]);
+
+impl StrkeyPreAuthTx {
+    pub fn to_string(&self) -> String {
+        encode(version::PRE_AUTH_TX, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        Ok(Self(payload.try_into().map_err(|_| DecodeError::Invalid)?))
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::PRE_AUTH_TX => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeyHashX(pub [u8; 32]);
+
+impl StrkeyHashX {
+    pub fn to_string(&self) -> String {
+        encode(version::HASH_X, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        Ok(Self(payload.try_into().map_err(|_| DecodeError::Invalid)?))
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::HASH_X => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+/// Stores a signed payload ed25519 signer.
+///
+/// The payload must not have a size larger than u32::MAX.
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeySignedPayloadEd25519 {
+    pub ed25519: [u8; 32],
+    pub payload: Vec<u8>,
+}
+
+impl StrkeySignedPayloadEd25519 {
+    /// Returns the strkey string for the signed payload signer.
+    ///
+    /// ### Panics
+    ///
+    /// When the payload is larger than u32::MAX.
+    pub fn to_string(&self) -> String {
+        let inner_payload_len = self.payload.len();
+        let payload_len = 32 + 4 + inner_payload_len + (4 - inner_payload_len % 4) % 4;
+
+        let inner_payload_len_u32: u32 = inner_payload_len
+            .try_into()
+            .expect("payload length larger than u32::MAX");
```

Comment:
Would it be better to do this?
```suggestion
        if inner_payload_len < 1 || inner_payload_len > 64 {
            panic!("payload must have a size between 1 and 64");
        }
```

Reply:
Payload length can be zero, there's restriction on it. There technically isn't a restriction on the max being 64 either, the max is u32::MAX as constrained by the 4-byte length field. But in practice only 64-byte strkeys are usable on network, so I think we should add the > 64 check, although not here. A check like that is more likely to occur and we should add the check at the point of building the object, not rendering to string. Let's do that in a follow up PR.